bin/remote-prover/src/utils.rs

```diff
@@ -1,52 +1,124 @@
+    session.write_response_body(Some(grpc_message.into()), false).await?;
+
+    // Send trailers with gRPC status
+    let mut trailers = HeaderMap::new();
+    let status_value = grpc_status.unwrap_or("0"); // Default to "0" (OK)
+    trailers.insert(
+        "grpc-status",
+        status_value.parse().map_err(|_| Error::new(ErrorType::InternalError))?,
+    );
+    session.write_response_trailers(trailers).await?;
+
+    Ok(true)
+}
+
+/// Write a gRPC error response to a Pingora session
+///
+/// This helper function creates a proper gRPC error response with the specified
+/// status code and error message.
+pub async fn write_grpc_error_to_session(
     session: &mut Session,
+    grpc_status: Code,
+    error_message: &str,
 ) -> pingora_core::Result<bool> {
-    // Set grpc-message header to "Too many requests in the queue"
-    // This is meant to be used by a Tonic interceptor to return a gRPC error
-    let mut header = ResponseHeader::build(503, None)?;
-    header.insert_header("grpc-message", "Too many requests in the queue".to_string())?;
-    header.insert_header("grpc-status", RESOURCE_EXHAUSTED_CODE)?;
+    // Create gRPC response headers (always HTTP 200 for gRPC)
+    let mut header = ResponseHeader::build(200, None)?;
+    header.insert_header("content-type", "application/grpc".to_string())?;
+
     session.set_keepalive(None);
-    session.write_response_header(Box::new(header.clone()), true).await?;
+    session.write_response_header(Box::new(header), false).await?;
 
-    let mut error = Error::new(ErrorType::HTTPStatus(503))
-        .more_context("Too many requests in the queue")
-        .into_in();
-    error.set_cause("Too many requests in the queue");
+    // gRPC errors don't have a body, just headers and trailers
+    session.write_response_body(None, false).await?;
 
-    session.write_response_header(Box::new(header), false).await?;
+    // Send trailers with gRPC status and error message
+    let mut trailers = HeaderMap::new();
+    let status_code = (grpc_status as i32).to_string();
+    trailers.insert(
+        "grpc-status",
+        status_code.parse().map_err(|_| Error::new(ErrorType::InternalError))?,
+    );
+
+    trailers.insert(
+        "grpc-message",
+        error_message.parse().map_err(|_| Error::new(ErrorType::InternalError))?,
+    );
+
+    session.write_response_trailers(trailers).await?;
+
+    Ok(true)
+}
 
+/// Create a gRPC `RESOURCE_EXHAUSTED` response for a full queue
+pub(crate) async fn create_queue_full_response(
+    session: &mut Session,
+) -> pingora_core::Result<bool> {
     // Increment the queue drop count metric
     QUEUE_DROP_COUNT.inc();
 
-    Err(error)
+    // Use our helper function to create a proper gRPC error response
+    write_grpc_error_to_session(session, Code::ResourceExhausted, "Too many requests in the queue")
+        .await
 }
 
-/// Create a 429 response for too many requests
+/// Create a gRPC `RESOURCE_EXHAUSTED` response for rate limiting
```

Comment:
I'm actually a bit unsure about moving rate-limiting away from http. As a general rule, 429 is well understood and _should_ usually include the backoff time i.e. when the client may retry without being rejected.

I guess I'm also surprised `pingora` doesn't include this itself somehow?

Reply:
Is this a global rate limit, or limit per client connection?

Reply:
I basically copied what pingora describes as [rate limiter](https://github.com/cloudflare/pingora/blob/main/docs/user_guide/rate_limiter.md), counting by client address.

Reply:
Pingora provides the `Rate` struct, which tracks events in a certain period of time.